[FIRRTL] Add module inlining pass

youngar opened the pull request:
The module inlining pass handles module flattening and instance
inlining.  Module flattening is recursively inlining every instance in a
module.  Module inlining is inlining only the instances of  modules
which are marked for inlining.

lattner commented on lib/Dialect/FIRRTL/Transforms/ModuleInliner.cpp:
```diff
@@ -0,0 +1,394 @@
+
+/// If this operation or any child operation has a name, add the prefix to that
+/// operation's name.
+static void rename(const Twine &prefix, Operation *op) {
```
The pervasive use of twine arguments here makes me concerned about dangling pointer bugs that happen due to twine temporaries.  Would you lose much by having these take StringRef prefixes instead?

youngar replied:
I had a bug with this when I was working on it, took me a while to find it even though I knew what the issue was going to be.  I have updated it to use StringRefs :)

lattner commented on lib/Dialect/FIRRTL/Transforms/ModuleInliner.cpp:
```diff
@@ -0,0 +1,394 @@
+  llvm::TypeSwitch<Operation *>(op)
+      .Case<CMemOp, InstanceOp, MemOp, MemoryPortOp, NodeOp, RegOp, RegResetOp,
+            SMemOp, WireOp>([&](auto op) {
+        // TODO: one day names won't be optional
```
Is there a bug tracking this? Seems like a high priority for the current sprint to get these representational issues sorted.

youngar replied:
This has been mostly fixed up in issue https://github.com/llvm/circt/issues/856.  There is one more change I want to make, and have opened a PR here: https://github.com/llvm/circt/pull/964.  (this PR is based on top of that change)

lattner commented on lib/Dialect/FIRRTL/Transforms/ModuleInliner.cpp:
```diff
@@ -0,0 +1,394 @@
+        // TODO: one day names won't be optional
+        if (auto name = op.nameAttr()) {
+          std::string newName = (prefix + name.getValue()).str();
+          op.nameAttr(StringAttr::get(op.getContext(), newName));
```
StringAttr::get is the thing that should take a Twine! :-)

lattner commented on lib/Dialect/FIRRTL/Transforms/ModuleInliner.cpp:
```diff
@@ -0,0 +1,394 @@
+/// Clone an operation, mapping used values and results with the mapper, and
+/// apply the prefix to the name of the operation. This will clone to the
+/// insert point of the builder.
+static void cloneAndRename(const Twine &prefix, OpBuilder b,
```
I'd recommend passing the OpBuilder by reference here.

youngar replied:
Fixed!

lattner commented on lib/Dialect/FIRRTL/Transforms/ModuleInliner.cpp:
```diff
@@ -0,0 +1,394 @@
+                                          FModuleOp target) {
+  SmallVector<Value> wires;
+  auto portInfo = target.getPorts();
+  for (unsigned i = 0; i < target.getNumArguments(); ++i) {
```
nit, no need to evaluate getNumArguments() each time through the loop

youngar replied:
Fixed x4

lattner commented on lib/Dialect/FIRRTL/Transforms/ModuleInliner.cpp:
```diff
@@ -0,0 +1,394 @@
+static void mapResultsToWires(BlockAndValueMapping &mapper,
+                              SmallVectorImpl<Value> &wires,
+                              InstanceOp instance) {
+  for (unsigned i = 0; i < instance.getNumResults(); ++i) {
```
likewise

lattner commented on lib/Dialect/FIRRTL/Transforms/ModuleInliner.cpp:
```diff
@@ -0,0 +1,394 @@
+    auto instanceName = instance.nameAttr().getValue();
+    auto nestedPrefix = (instanceName + "_").str();
+    auto wires = mapPortsToWires(nestedPrefix, b, mapper, target);
+    for (unsigned i = 0; i < instance.getNumResults(); ++i)
```
same nit

lattner commented on lib/Dialect/FIRRTL/Transforms/ModuleInliner.cpp:
```diff
@@ -0,0 +1,394 @@
+    auto instanceName = instance.nameAttr().getValue();
+    auto nestedPrefix = (instanceName + "_").str();
+    auto wires = mapPortsToWires(nestedPrefix, b, mapper, target);
+    for (unsigned i = 0; i < instance.getNumResults(); ++i)
```
Here and in a few other places too

lattner commented on lib/Dialect/FIRRTL/Transforms/ModuleInliner.cpp:
```diff
@@ -0,0 +1,394 @@
+
+void Inliner::run() {
+  // If the top module is not a regular module, there is nothing to do.
+  auto topModule = dyn_cast<FModuleOp>(symbolTable.lookup(circuit.name()));
```
CircuitOp has a helper for this, plz use it.

youngar replied:
Fixed!  I was worried that the lookup function would construct its own symbol table, but looks like it just does a linear scan through IR.

lattner commented on lib/Dialect/FIRRTL/Transforms/ModuleInliner.cpp:
```diff
@@ -0,0 +1,394 @@
+  if (topModule) {
+    worklist.push_back(topModule);
+  }
+  liveModules.insert(topModule);
```
Is this correct in the null case?

youngar replied:
This was busted, I got confused at some point and thought you couldn't have an ExtModule as the main module, and I actually meant to delete this check.

Apparently that isn't remotely correct and I have fixed this check and added a test case where the top module is an extmodule.

lattner commented on lib/Dialect/FIRRTL/Transforms/ModuleInliner.cpp:
```diff
@@ -0,0 +1,394 @@
+
+  // Delete all unreferenced modules.
+  for (auto &op : llvm::make_early_inc_range(*circuit.getBody())) {
+    if (isa<FModuleOp, FExtModuleOp>(op) && !liveModules.count(&op)) {
```
Circuit can only contain modules now that the terminator is gone, so you can delete the isa check (or change it into an assert)

lattner reviewed (approved):
Looking great!

lattner commented:
Hey @youngar what's your plan with this patch?  Should I re-review it or something?  Why not land it?

youngar commented:
@lattner I should have probably landed it a while ago.  I was just waiting until all the unresolved issues were solved and the last one remaining was flip canonicalization. I think I should do one more pass over how the wires are created for input and output ports and then i'll land it.

lattner commented:
Aha got it.  If you'd like to wait for @seldridge 's output patch to land that is also fine.  I didn't realize there was a dependency, thanks!

youngar commented:
I added a test which inlined a module with bundle parameters, and added some code to strip the outer flip from output ports. I think this is a good time to merge this PR since the output port semantics are settled, we are just changing the representation.
